engine/consensus/approvals/aggregated_signatures.go

```diff
@@ -43,10 +45,15 @@ func (as *AggregatedSignatures) HasSignature(chunkIndex uint64) bool {
 
 // Collect returns array with aggregated signature for each chunk
 func (as *AggregatedSignatures) Collect() []flow.AggregatedSignature {
-	aggregatedSigs := make([]flow.AggregatedSignature, len(as.signatures))
-
 	as.lock.RLock()
 	defer as.lock.RUnlock()
+
+	// if there aren't signatures for each chunk we can't collect signatures
+	if uint64(len(as.signatures)) < as.numberOfChunks {
+		return nil
```

Comment:
Would it be better to return some error here instead of nil?

Reply:
Is a [seal](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L105) still valid with no aggregated signature?

Reply:
I don't think that we should error here, being no-op should be sufficient since it's more a check to support the invariant of having only valid indexes. Actual checks for chunk index range are performed way earlier before getting to this part.

> Is a [seal](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L105) still valid with no aggregated signature?

In terms of signatures count seal is valid iff it has a signature for every chunk. 
But since we have emergency sealing nodes can [skip](https://github.com/onflow/flow-go/blob/683fdd1209a87226e95aab8ec740ac10a2447608/module/validation/seal_validator.go#L249) validating this condition.

Reply:
returning `nil` here: this would essentially introduce two different conventions: 
1. if we require approvals, we expect that there is an aggregated signature for each chunk 
2. if we don't require approvals, there are no aggregated signatures 

My preference would be to stick to a uniform convention that naturally covers both cases:
  * the seal should contain an aggregated signature for each chunk (in all cases).
  * for each chunk, the number of approvers must be larger or equal to `requiredApprovalsForSealVerification`. I.e. when we don't require any approvals, we include an empty `AggregatedSignatures`.

Our current implementation of the `sealValidator` accepts both conventions: https://github.com/onflow/flow-go/blob/77cd71b7930a2016fa09a1b79a2afa459b237903/module/validation/seal_validator.go#L249-L258 Nevertheless, my preference would be (on the generating side) to stick with one uniform convention that includes the degenerate edge case. Thereby, the `sealValidator` is flexible in what it accepts, but we generate the seals according to a single convention.